code/headers/hardware_usart.hpp

```diff
@@ -1,175 +1,278 @@
 #pragma once
+/**
+ * @file hardware_usart.hpp
+ * @author Patrick Dekker
+ * @brief Hardware implementation for the usart
+ * @version 0.1
+ * @date 2019-05-24
+ *
+ * @copyright Copyright (c) 2019
+ *
+ */
 
 #include <hwlib.hpp>
-#include <queue.hpp>
+#include <ringbuffer.hpp>
+#include <type_traits>
+
 #include <usart_connection.hpp>
 
 namespace r2d2::usart {
+    namespace detail::pio {
+        struct periph_a {};
+        struct periph_b {};
 
-    enum class usart_ports {
-        uart1 = 0,
-        uart2,
-        uart3,
-        // only for determining how many uart ports there are
-        UART_SIZE
-    };
+        struct pioa {};
+        struct piod {};
+    }; // namespace detail::pio
 
-    enum class peripheral {
-        peripheral_a,
-        peripheral_b
+    struct usart0 {
```

Comment:
I should probably rename this back to the name on the port that the arduino due has.

Reply:
Added usart port 2. This port doesnt have a TX/RX label. So I added some doxygen to comment what usart is using what pin